ftd2xx/aio/transports.py

```diff
@@ -0,0 +1,453 @@
+        except DeviceError:
+            pass
+
+        try:
+            self._protocol.connection_lost(exc)
+        finally:
+            self._post_connection_lost()
+
+    def _post_connection_lost(self):
+        self._ftd2xx.close()
+        self._ftd2xx = None
+        self._protocol = None
+        self._loop = None
+
+
+class FTD2xxReadTransport(FTD2xxBaseTransport, asyncio.ReadTransport):
+    """FTDI D2XX read-only transport."""
+
+    def __init__(self, loop, protocol_factory, ftd2xx_instance):
+        super().__init__(loop, protocol_factory, ftd2xx_instance)
+        self._max_read_size = 1024
+        self._has_reader = False
+        loop.call_soon(self._ensure_reader)
+
+    def _read_ready(self):
+        try:
+            data = self._ftd2xx.read(self._max_read_size)
+        except DeviceError as exc:
+            self._close(exc=exc)
+        else:
+            if data:
+                self._protocol.data_received(data)
+
+    def pause_reading(self):
+        """Pause the receiving end of the transport.
+        No data will be passed to the protocol’s data_received() method
+        until resume_reading() is called.
+        """
+        self._remove_reader()
+
+    def resume_reading(self):
+        """Resume the receiving end of the transport.
+        Incoming data will be passed to the protocol's data_received()
+        method until pause_reading() is called.
+        """
+        self._ensure_reader()
+
+    def _shutdown(self):
+        self._remove_reader()
+        super()._shutdown()
+
+    def _graceful_shutdown(self) -> bool:
+        if super()._graceful_shutdown():
+            self._remove_reader()
+            return True
+        return False
+
+    def _call_connection_lost(self, exc):
+        assert not self._has_reader
+        super()._call_connection_lost(exc)
+
+    if sys.platform == "win32":
+
+        def _poll_read(self):
+            if self._has_reader and not self._closing:
+                try:
+                    self._has_reader = self._loop.call_later(
+                        self._poll_wait_time, self._poll_read
+                    )
+                    # if self.serial.in_waiting:
+                    if True:  # TODO: ???
+                        self._read_ready()
+                except DeviceError as exc:
+                    self._fatal_error(exc, "Fatal write error on ftd2xx transport")
+
+        def _ensure_reader(self):
+            if not self._has_reader and not self._closing:
+                self._has_reader = self._loop.call_later(
+                    self._poll_wait_time, self._poll_read
+                )
+
+        def _remove_reader(self):
+            if self._has_reader:
+                self._has_reader.cancel()
+            self._has_reader = False
+
+    else:
+
+        def _ensure_reader(self):
+            if (not self._has_reader) and (not self._closing):
+                self._loop.add_reader(self._ftd2xx.fileno(), self._read_ready)
```

Comment:
@snmishra Can you even get a file descriptor? Maybe the polling method needs to be used on linux as well.

Reply:
Perhaps we could use setEventNotification somehow? I have not used before so I don't know how it works.

Reply:
I don't have any idea either. FTD2XX class doesn't have file descriptors for sure.

Reply:
@krister-ts Thanks for continued work on this. I don't have bandwidth right now. Hopefully I can check again on Sunday. Sorry, I'll be silent for a bit.